REP-6196 Disallowing inline functions in policy path expressions

dmnjohns opened the pull request:
The net effect of this change is to allow dynamic function application (e.g., `$map($key)` in place of `map:get($map, $key)`) and to disallow inline function usage (and, consequently, infinite recursion).

To be clear, inline function definitions were not previously allowed in policy path expressions since the `XPathParser` would have thrown an error when parsing. This change should make that more clear, though. Map functions would have worked, but not if function call syntax was used.

**EDIT:** I've updated this PR to include refactoring for the validate policy suite.

dmnjohns commented on core/src/main/scala/com/rackspace/identity/components/ValidateXPathFunction.scala:
```diff
@@ -122,6 +129,7 @@ class ValidateXPathFunction(conf: Configuration) extends ExtensionFunction {
   //
   private def parseXPath(expression: String, context: StaticContext): Expression = {
     val xpathParser = new XPathParser
+    xpathParser.setParserExtension(new HofParserExtension)
```
Based on an error message I was seeing when I did not set the parser extension, I was concerned that this would force the usage of Saxon PE/EE. However, I was able to run the tests just fine, even when I moved my Saxon license file to a random location. I also looked at the Hof extension code and did not see any license checks, so I think we should be alright.

dmnjohns commented on core/src/main/scala/com/rackspace/identity/components/ValidateXPathFunction.scala:
```diff
@@ -131,11 +139,15 @@ object ValidateXPathFunction {
   private final val ENVIRONMENT_VARIABLE_FUNCTION_NAME = "environment-variable"
+  private final val FUNCTION_LOOKUP_FUNCTION_NAME = "function-lookup"
+  private final val LOAD_XQUERY_MODULE_FUNCTION_NAME = "load-xquery-module"
+  private final val TRANSFORM_FUNCTION_NAME = "transform"
```
It may not be apparent since tests are generated based on file locations, but even without the explicit blacklisting here, these functions are not allowed. The reason is described in a todo on the tests. I've left the blacklisting in place so that, if/when the `XPathParser` class supports HOFs, it will just be a matter of moving the tests to where they actually belong.

More details are provided in the todo and any corresponding PR comments.

dmnjohns commented on core/src/test/scala/com/rackspace/identity/components/ValidatePolicySuite.scala:
```diff
@@ -20,133 +20,125 @@ import javax.xml.transform.stream.StreamSource
+  // TODO: "Higher-order functions" are not currently supported by Saxon's XPathParser.
+  // TODO: When they are, this category should be merged into the blacklisted category.
+  // TODO: Additionally, "safe" higher-order functions (i.e., those duplicated in the "valid/hof" directory) should be
+  // TODO: removed from the invalid test category.
```
From what I can tell, Saxon 9.8.0-4's `XPathParser` does not support parsing of most [higher-order functions](https://www.w3.org/TR/xpath-functions-31/#higher-order-functions). Those functions are considered missing, and there is a line in the source (in the `getMissingFunctionExplanation` method) that "explains" the missing functions with this message:
> Higher-order functions are not available in this Configuration

Even running the tests with a valid Saxon license does not enable parsing those functions with the `XPathParser`.

However, since the intent was to blacklist certain higher-order functions and allow the rest, but allowing the rest was considered a nice to have, it seems like just saying all higher-order functions are blacklisted at the moment is an acceptable solution. So here that is.

dmnjohns commented:
I'll fix the Scalastyle violations as soon as I have a moment.

I wrote some additional tests (store locally at the moment) that exhibit a different issue, though. The XPath expression we've written to validate user-provided XPath expressions does not handle multiple lines well. I remember dealing with that for other processing, but it seems we missed it for validation.

RackerWilliams commented:
Understood.

Quick question:  Are you setting things up so that HOF support improves in the XPathParser -- we will somehow know that that's happened -- maybe via a test failure.

dmnjohns commented:
Well yes, the HOF tests I have added will start to fail when support is brought in. I also added that `todo` to kind of describe what's going on. I did not write a specific test with a message stating that when it fails, these HOFs are supported and so the tests should be recategorized. I thought the tests and `todo` did a good enough job of explaining what is going on, but I can break that into a specific test if you prefer.

RackerWilliams commented on core/src/main/scala/com/rackspace/identity/components/AttributeMapper.scala:
```diff
@@ -127,7 +127,7 @@ object AttributeMapper {
           We store the result of analyze-string, which should give us all the capture groups!
       :)
-      $localXPathMatches := for $value in $localValuesWithXPaths return analyze-string($value, '\{Pts?\((.*?)\)\}'),
+      $localXPathMatches := for $value in $localValuesWithXPaths return analyze-string($value, '\{Pts?\((.*?)\)\}', 'ms'),
```
Ah, I was afraid I'd miss that somewhere!

RackerWilliams commented on core/src/main/scala/com/rackspace/identity/components/ValidateXPathFunction.scala:
```diff
@@ -147,18 +159,25 @@ object ValidateXPathFunction {
+    FUNCTION_LOOKUP_FUNCTION_NAME,
+    LOAD_XQUERY_MODULE_FUNCTION_NAME,
+    TRANSFORM_FUNCTION_NAME,
     UNPARSED_TEXT_AVAILABLE_FUNCTION_NAME,
```
Missing `apply` according to this : https://www.w3.org/TR/xpath-functions-31/#func-apply  apply is in the fn namespace.

RackerWilliams commented on core/src/main/scala/com/rackspace/identity/components/ValidateXPathFunction.scala:
```diff
@@ -147,18 +159,25 @@ object ValidateXPathFunction {
   )
+  private final val EXCEPTION_FUNCTION_MAP = Map(
+    // Enables dynamic function application
+    SAXON -> APPLY_FUNCTION_NAME
```
Is this a different apply -- saxon specific extension?

dmnjohns replied:
Yes. I believe this is Saxon's dynamic function application function. From what I can tell, when a user declares an inline function, Saxon actually gives that function a (semi-random) name. To apply that function, Saxon uses this `apply` function which lives in their namespace.

Interestingly enough, this function is not documented with the rest of the functions in the [http://www.saxonica.com/html/documentation/functions/saxon/](Saxon documentation), which leads me to believe that this function is meant primarily for internal use.

RackerWilliams replied:
Wow, it's scary that there's an undocumented function that has that much power.  Good thing we're blocking based on namespace as well.

RackerWilliams commented on core/src/test/resources/tests/validate-policy-tests/invalid/blacklisted-hof/function-lookup-s.json:
```diff
@@ -0,0 +1,21 @@
+        ],
+        "version":"RAX-1",
+        "namespaces": {
+            "xs": "http://www.w3.org/2001/XMLSchema"
```
I'm not positive, but I think you need to define fn here.

dmnjohns replied:
Interestingly enough, that does not seem to be the (at least when it comes to `XPathParser`). When I was writing the tests, I noticed that even without prefixing functions in the conventional `fn` namespace, Saxon parsed those functions correctly. As in, Saxon consider those functions in be in the `fn` namespace. Perhaps Saxon assumes non-namespaced functions must be in the `fn` namespace?

I can be more explicit, if you'd like.

RackerWilliams replied:
I guess it's okay, the namespace must be added by default.

RackerWilliams commented on core/src/main/scala/com/rackspace/identity/components/ValidateXPathFunction.scala:
```diff
@@ -147,18 +159,25 @@ object ValidateXPathFunction {
+    FUNCTION_LOOKUP_FUNCTION_NAME,
+    LOAD_XQUERY_MODULE_FUNCTION_NAME,
+    TRANSFORM_FUNCTION_NAME,
     UNPARSED_TEXT_AVAILABLE_FUNCTION_NAME,
```
Shouldn't `APPLY_FUNCTION_NAME` be here.  see https://www.w3.org/TR/xpath-functions-31/#func-apply

dmnjohns replied:
Oops, not sure how I added the constant for the function name, but didn't stick it in this set. I guess tests still passed since it isn't supported yet, but it should still be here. Will fix.

dmnjohns replied:
Ah. It's early, so maybe I forgot. I think I intentionally allow `apply`. This was the function we talked about a bit over Zoom, right? It shouldn't be dangerous to allow because you have to hand it a reference to a function. Well, if that reference is to a user-defined function, then the `XPathParser` should parse a `UserFunctionReference` due to the Saxon magic that I referred to in another comment. In that case, the expression should be considered invalid.

I actually broke this out into two test cases: `apply` and `recursive-apply`. I expect `apply` to validate when passed a function like `concat`, but not when passed a user-defined (in this case, recursive) function. Of course, since HOF support is not there yet, both of these cases currently fail validation.

RackerWilliams replied:
Ah that makes sense!  Thanks.

RackerWilliams commented on core/src/main/scala/com/rackspace/identity/components/ValidateXPathFunction.scala:
```diff
@@ -147,18 +159,25 @@ object ValidateXPathFunction {
   )
+  private final val EXCEPTION_FUNCTION_MAP = Map(
+    // Enables dynamic function application
+    SAXON -> APPLY_FUNCTION_NAME
```
I'm assuming there is a Saxon extension that also implements apply?
